std-bits/base/src/main/java/org/enso/base/numeric/Decimal_Utils.java

```diff
@@ -0,0 +1,88 @@
+package org.enso.base.numeric;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.math.MathContext;
+
+/** Utils for the Enso Decmial type. */
+public class Decimal_Utils {
+  private static final BigDecimal MIN_LONG_BIGDECIMAL = BigDecimal.valueOf(Long.MIN_VALUE);
+  private static final BigDecimal MAX_LONG_BIGDECIMAL = BigDecimal.valueOf(Long.MAX_VALUE);
+
+  public static BigDecimal fromString(String s) {
+    return new BigDecimal(s);
+  }
+
+  public static Conversion_Result fromString(String s, MathContext mc) {
+    BigDecimal bd = new BigDecimal(s, mc);
+    BigDecimal withoutMC = new BigDecimal(s);
+    return new Conversion_Result(bd, bd.compareTo(withoutMC) != 0);
+  }
+
+  public static BigDecimal fromInteger(Object o) {
+    if (o instanceof Long l) {
+      // According to the BigInteger Javadocs, valueOf is preferred "because it
+      // allows for reuse of frequently used BigDecimal values".
+      return BigDecimal.valueOf(l);
+    } else if (o instanceof BigInteger bi) {
+      return new BigDecimal(bi);
+    } else {
+      throw new IllegalArgumentException("Input must be Long or BigInteger");
+    }
+  }
+
+  public static Conversion_Result fromInteger(Object o, MathContext mc) {
+    if (o instanceof Long l) {
+      BigDecimal bd = new BigDecimal(l, mc);
+      BigDecimal withoutMC = new BigDecimal(l);
+      long backToLong = bd.longValue();
+      return new Conversion_Result(bd, bd.compareTo(withoutMC) != 0);
+    } else if (o instanceof BigInteger bi) {
+      BigDecimal bd = new BigDecimal(bi, mc);
+      BigDecimal withoutMC = new BigDecimal(bi);
+      BigInteger backToBigInteger = bd.toBigInteger();
+      return new Conversion_Result(bd, bd.compareTo(withoutMC) != 0);
+    } else {
+      throw new IllegalArgumentException("Input must be Long or BigInteger");
+    }
+  }
+
+  public static BigDecimal fromFloat(Double d) {
+    // According to the BigInteger Javadocs, valueOf is preferred because "the
+    // value returned is equal to that resulting from constructing a BigDecimal
+    // from the result of using Double.toString(double)."
+    return BigDecimal.valueOf(d);
+  }
+
+  public static BigDecimal fromFloat(Double d, MathContext mc) {
+    // We do not check for precision loss here because we always attach a
+    // warning when converting from float.
+    return new BigDecimal(d, mc);
```

Comment:
Then maybe we should always return a `Conversion_Result` with `precisionLoss=true` here?

It feels a bit safer IMHO if we handle all kinds of numeric types in the same way (always returning the `Conversion_Result`), then on the Enso side we could have more unified handling.

It also feels less likely to forget to add the 'always added' warning in case of floats, if the `Conversion_Result` is being returned.

Reply:
Done